lib/src/main/kotlin/Node.kt

```diff
@@ -1,3 +1,17 @@
-internal open class Node<TChild>(parent: NodeBase?) : NodeBase(parent) where TChild : NodeBase {
-    var children = mutableListOf<TChild>()
+abstract class Node<TAction, TSelf>(
+    val parent: TSelf?,
```

Comment:
Not sure how TSelf in `val parent: TSelf?` is utilized here.

Reply:
Aha, this is a great question. Kinda hard to explain but it's a limitation of the typing system in Kotlin. See https://medium.com/@jerzy.chalupski/emulating-self-types-in-kotlin-d64fe8ea2e62. I can't use extension methods here though since each implementation of Node will potentially have a different addChild/getChildren.

Reply:
- Extends from generic BaseClass
- We wish to return our own object type
- Ideally done via Extension methods, but we want the methods to be overwritten by subclasses. But extension methods cannot be overwritten, and has same behaviour.
- So we use self.type

API for extending solver cleaner, rather than doing a bunch of casting.

Reply:
Not the cleanest, but a valid solution.

Reply:
Otherwise requires typecast everywhere.